python/taichi/lang/transformer.py

```diff
@@ -642,7 +641,7 @@ def visit_FunctionDef(self, node):
             # Transform as func (all parameters passed by value)
             arg_decls = []
             for i, arg in enumerate(args.args):
-                if i == 0 and self.is_classfunc:
+                if isinstance(self.func.arguments[i], ti.template):
                     continue
```

Comment:
```suggestion
                # Directly pass in template arguments,
                # such as class instances ("self"), tensors, SNodes, etc.
                if isinstance(self.func.arguments[i], ti.template):
                    continue
                # Create a copy for non-template arguments,
                # so that they are passed by value.
```

Reply:
Aliright thanks! Yes, both tests are passing. Although the second one passes even without making x a template. I assume its related to the hack https://github.com/taichi-dev/taichi/commit/0139195702234fd77d6c33f9eedb87255ab98690#r39384288  @archibate  was talking about, which I think we can get rid of at some point.